drivers/rn2xx3/rn2xx3_internal.c

```diff
@@ -0,0 +1,355 @@
+            puts("IDLE");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_JOIN:
+            puts("JOIN");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_RX_MESSAGE:
+            puts("RX MSG");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_RX_PORT:
+            puts("RX PORT");
+            break;
+
+        case RN2XX3_INT_STATE_MAC_TX:
+            puts("TX");
+            break;
+
+        case RN2XX3_INT_STATE_RESET:
+            puts("RESET");
+            break;
+
+        case RN2XX3_INT_STATE_SLEEP:
+            puts("SLEEP");
+            break;
+
+        default:
+            puts("UNKNOWN");
+            break;
+    }
+#endif
+
+    dev->int_state = state;
+}
+
+int rn2xx3_write_cmd(rn2xx3_t *dev)
+{
+    int ret;
+    DEBUG("[rn2xx3] CMD: %s\n", dev->cmd_buf);
+
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_CMD);
+
+    mutex_lock(&(dev->cmd_lock));
+    _uart_write(dev, dev->cmd_buf);
+    _uart_write(dev, closing_seq);
+
+    ret = rn2xx3_wait_response(dev);
+    if (ret == RN2XX3_TIMEOUT) {
+        mutex_unlock(&(dev->cmd_lock));
+        return RN2XX3_TIMEOUT;
+    }
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    ret = rn2xx3_process_response(dev);
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_IDLE);
+
+    DEBUG("RET: %d, RESP: %s\n", ret, dev->resp_buf);
+
+    return ret;
+}
+
+int rn2xx3_write_cmd_no_wait(rn2xx3_t *dev)
+{
+    DEBUG("[rn2xx3] CMD: %s\n", dev->cmd_buf);
+
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    mutex_lock(&(dev->cmd_lock));
+    _uart_write(dev, dev->cmd_buf);
+    _uart_write(dev, closing_seq);
+
+    DEBUG("[rn2xx3] RET: %s\n", dev->resp_buf);
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    return rn2xx3_process_response(dev);
+}
+
+int rn2xx3_wait_response(rn2xx3_t *dev)
+{
+    if (_wait_reply(dev, RESP_TIMEOUT_SEC)) {
+        return RN2XX3_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] RESP: %s\n", dev->resp_buf);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_wait_reply(rn2xx3_t *dev, uint8_t timeout)
+{
+    if (_wait_reply(dev, timeout)) {
+        return RN2XX3_REPLY_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] REPLY: %s\n", dev->resp_buf);
+
+    return rn2xx3_process_reply(dev);
+}
+
+void rn2xx3_cmd_start(rn2xx3_t *dev)
+{
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_CMD);
+    DEBUG("[rn2xx3] CMD: %s", dev->cmd_buf);
+    mutex_lock(&(dev->cmd_lock));
+    _uart_write(dev, dev->cmd_buf);
+}
+
+void rn2xx3_cmd_append(rn2xx3_t *dev, const uint8_t *payload, uint8_t payload_len)
+{
+    char payload_str[3];
+    for (unsigned i = 0; i < payload_len; i++) {
+        rn2xx3_bytes_to_hex(&payload[i], payload_str, 1);
+        uart_write(dev->p.uart, (uint8_t*)payload_str, 2);
```

Comment:
Why not to use `_uart_write` also here?

Reply:
because payload_str has a length of 3 and I just want to write 2 chars.